Interactivity API: Fix directives in tags whose closing tag is not visited not working

SantosGuillamot opened the pull request:
Trac ticket: https://core.trac.wordpress.org/ticket/60746

Right now, if I add directives to tags with closing tag that is not visited by the tag processor, it breaks the whole server-side rendering processing.

This is the list of those tags: https://github.com/WordPress/wordpress-develop/blob/186eeafb6888431fdf445b2abf2036f0606edcd5/src/wp-includes/interactivity-api/class-wp-interactivity-api-directives-processor.php#L25-L34

For example, if I try to add a directive to an iframe, it breaks the processing.

Co-authored by [@cbravobernal](https://profiles.wordpress.org/cbravobernal)

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

mukeshpanchal27 commented on src/wp-includes/interactivity-api/class-wp-interactivity-api.php:
```diff
@@ -281,9 +283,9 @@ private function process_directives_args( string $html, array &$context_stack, a
 
 					/*
-					 * If this tag will visit its closer tag, it adds it to the tag stack
-					 * so it can process its closing tag and check for unbalanced tags.
-					 */
+					* If this tag will visit its closer tag, it adds it to the tag stack
+					* so it can process its closing tag and check for unbalanced tags.
+					*/
```
Should be revert.

SantosGuillamot replied:
Thank you! Done in [this commit](https://github.com/WordPress/wordpress-develop/pull/6247/commits/f2dd8b64448cb670665e4b1b8bc417343f7db435).

swissspidy commented:
cc for review: @luisherranz, maybe @westonruter too

westonruter commented on tests/phpunit/tests/interactivity-api/wpInteractivityAPIFunctions.php:
```diff
@@ -390,4 +390,31 @@ public function test_wp_interactivity_data_wp_context_with_json_flags() {
+	 *
+	 * @covers ::wp_interactivity_process_directives_of_interactive_blocks
+	 */
+	public function test_process_directives_in_tags_that_dont_visit_closer_tag() {
```
Could this test be expanded (or another one added) which test that the absence of the new `if` condition does indeed cause breaking the whole server-side rendering processing as described?

SantosGuillamot replied:
Not sure if I understand what you mean. The existing test already checks that the SSR doesn't break when there is an iframe. The same test was failing without the final solution.

By the way, I've reverted the `if` condition because it is solved by another PR as explained [here](https://github.com/WordPress/wordpress-develop/pull/6247#issuecomment-1989598180).

westonruter replied:
I thought that there were wider breakages than just with a single element. So that's what I thought would be good to test.

dmsnell commented:
After a lot of debugging I'm wondering if the IFRAME is a _red herring_ and the problem is that we have directive processors that don't handle elements without closing tags. For example, I added a test which adds context on an IMG element and it also failed to process in the same way that the test added here fails.

```html
<div data-wp-context='{"text": "outer"}'>
<img data-wp-context='{"text": "inner"}'>
<p data-wp-text="context.text"></p>
</div>
```

and then in the test I call `do_blocks()` on the HTML without doing anything else

![Screenshot 2024-03-11 at 12 25 40 PM](https://github.com/WordPress/wordpress-develop/assets/5431237/1b4d5425-8da6-463f-be58-c65726370045)


Or these tests simply aren't running the server directive processor because when I output the input `$html` I'm not seeing that in the logs, only for two of the tests in the `wpInteractivityApiFunctions.php` suite - `_of_single_interactive_block()` and `_only_process_the_root_interactive_block`

dmsnell commented:
please ignore my last comment: I finally reproduced this and I think I can say that it's fixed already in #6120, which is supposed to go in to 6.5 anyway, and this gives it even more reason to do so.

**what's happening?**

once the attribute was updated and `next_tag()` is called, the Tag Processor applies the changes and then re-parses content. it wasn't doing that properly once it started scanning all tokens, because it was short-circuiting some logic to avoid sub-class methods from being called (e.g. you don't want `WP_HTML_Processor::next_token()` to be called and then mess with the HTML stack of open elements).

unfortunately this short-circuit was programmed with the old assumption that everything is a tag, and we missed updating it earlier in the cycle for all tokens, including the handling of special tokens.

in #6120 the Tag Processor creates its own internal `base_class_next_token()` that it calls so that all of the logic is repeated after applying updates and seeking. this resolves the issue here because it then is able to trap the start and end of the `iframe` tag as one token.

dmsnell commented:
This branch, without the proposed fix, but with the deferred updates PR passes the tests.
https://github.com/WordPress/wordpress-develop/pull/6250

SantosGuillamot committed:
Add test for iframe

SantosGuillamot committed:
Fix comment spacing

SantosGuillamot committed:
Fix tags with closing tag not visited

SantosGuillamot committed:
Revert "Fix tags with closing tag not visited"

This reverts commit d11ebcecdd60b1c9cfd5ab68f1f4e0f393ce52cd.

SantosGuillamot commented:
Thanks a lot for taking a deeper look at it, @dmsnell ☺️ 

I've rebased `trunk` and reverted the proposed fix and I can confirm that the test passes now. I'll leave this pull request only with the new test to ensure it keeps working as expected in the future.

dmsnell commented:
Merged in [[57822]](https://core.trac.wordpress.org/changeset/57822)
https://github.com/wordpress/wordpress-develop/commit/b1b97f96727f137210b942f00f85fc6449033d1d
